src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs

```diff
@@ -2027,19 +2210,16 @@ private static BigInteger Subtract(uint[]? leftBits, int leftSign, uint[]? right
                 }
             }
 
-            int zl = xd.Length - digitShift + (trackSignBit ? 1: 0);
-            uint[]? zdArray = null;
-            Span<uint> zd = stackalloc uint[0];
-            if (zl > 0)
-            {
-                zd = zl <= StackallocUInt32Limit ?
-                    stackalloc uint[StackallocUInt32Limit].Slice(0, zl) :
-                    zdArray = new uint[zl];
-            }
+            uint[]? zdFromPool = null;
+            int zl = Math.Max(xl - digitShift, 0) + (trackSignBit ? 1 : 0);
+            Span<uint> zd = ((uint)zl <= BigIntegerCalculator.StackAllocThreshold ?
+                            stackalloc uint[BigIntegerCalculator.StackAllocThreshold]
+                            : zdFromPool = ArrayPool<uint>.Shared.Rent(zl)).Slice(0, zl);
+            zd.Clear();
 
             if (smallShift == 0)
             {
-                for (int i = xd.Length - 1; i >= digitShift; i--)
+                for (int i = xl - 1; i >= digitShift; i--)
```

Comment:
Prefer `xd.Length` instead of `xl` here and for the remainder of the method. It's more readable and will actually produce slightly better codegen (more efficient register allocation).